docs/rfcs/2024-10-07-chat-infinite-scrolling.md

```diff
@@ -0,0 +1,322 @@
+# Chat infinite scrolling
+
+## Problem
+
+The current scroll-based data loading model, where new messages are loaded as users scroll upwards, introduces several usability challenges for the app:
+- The system cannot reliably open a chat at the first unread message on a given chat because data may not yet be loaded for that page.
+- Features like tapping on a quote to jump to older messages are not supported due to the absence of a stable mechanism to navigate across distant points in the chat history.
+- When users search for a message, the results are shown without surrounding context, severely limiting the usefullness of this feature
+
+## Solution
+
+Adopting a bi-directional infinite scroll system that fetches and displays chat data on demand as users scroll both upwards and downwards.
+
+This approach will directly impact all users by enabling:
+- **Seamless navigation:** Users can jump to any point in the chat history (e.g., first unread message or an older quoted message) without the need to load the entire conversation.
+- **Enhanced search experience:** Search results will be more useful by fetching both the matching messages and their surrounding context, allowing users to follow the flow of the conversation.
+
+At the same time, this system will:
+- Optimize memory usage.
+- Contain database fetch operations.
+- Maintain smooth scrolling.
+
+### Pagination integration:
+
+```
+fetch(scrollingUp, chatItemId)
+    if scrollingUp
+        fetchBefore(chatItemId)
+    else
+        fetchAfter(chatItemId)
+
+-- Another option would be if the api exposed a way to do before and after, possibly already supported since they are just separate query strings
```

Comment:
it is already supported, yes

Reply:
current approach to pagination in backend (based on ts, database id) is problematic though, see comments below - maybe we should rework backend simultaneously with this.
https://github.com/simplex-chat/simplex-chat/pull/3677#issuecomment-1889326757
https://github.com/simplex-chat/simplex-chat/pull/3661#issuecomment-1884824486

Reply:
For my understanding, is the major issue we face that IDs are not stable and therefore results of pagination may not be immutable? (Ex if I hit the same page twice totally different results may be retrieved)

Reply:
How does the issue manifest today for users? Duplicates? Missing messages?

Reply:
I guess we also have no way of knowing if a permanent delete happens, since id may be reused

Reply:
There's multiple issues arising out of it, the main reason for all of them is that IDs are not guaranteed to be monotonically increasing, because deleted records free up old IDs that can be reused.

So for a given item ID is stable, but sorting based on ID doesn't guarantee correct item order. So we sort on ts first*, then ID. Using ID for pagination cutoff is also incorrect (as done for contact and local chat queries currently; groups use ts instead - see `getDirectChat`, `getGroupChat`, `getLocalChat` pagination queries). For example, you search for items before item X, but it reused old ID - current query would incorrectly filter out items with greater IDs that really are before item X chronologically. At the very least contact and local queries should be reworked similar to group queries, but it wouldn't solve all problems.

* Using ts is also not perfect, depending on where it originated: server ts used for sorting received items in groups are rounded to second, so messages that have same ts on server may be incorrectly sorted on ID fallback; local ts is unreliable because it depends on time set on user device, which user can change arbitrarily. For example, if user sets device time forward 10 mins, in a group all sent messages are created with it, received messages are created with server time; conversation happens, and re-opening chat shows only sent items last (it looks like received items have disappeared unless you scroll up - it's not a conjecture, it practically happened to a user).

How it can manifest:
- possible chronologically incorrect item ordering / different item ordering on chat reload from one when you are inside chat and receiving events
- possible skips on pagination (even cutting off based on ts does not guarantee no skips I think in case items had same ts and were sorted on ID)
- messages fail to be marked as read (via range api)

What we need to solve all these issues is some counter tables with monotonically increasing counters per conversation, linking back to chat items.

Reply:
These are all arguments for causality-based sorting really.